internal/error.go

```diff
@@ -27,82 +27,98 @@ package internal
+		// handle activity errors (created via NewApplicationError() API)
+		if !applicationErr.NonRetryable() {
+			// manually retry activity
+		}
+		var detailMsg string // assuming activity return error by NewApplicationError("message", true, "string details")
+		applicationErr.Details(&detailMsg) // extract strong typed details
+
+		// handle activity errors (errors created other than using NewApplicationError() API)
+		switch err.OriginalType() {
+		case "CustomErrTypeA":
+			// handle CustomErrTypeA
+		case CustomErrTypeB:
+			// handle CustomErrTypeB
 		default:
 			// newer version of activity could return new errors that workflow was not aware of.
 		}
-	case *workflow.GenericError:
-		// handle generic error (errors created other than using NewCustomError() API)
-	case *workflow.CanceledError:
+	}
+
+	var canceledErr *CanceledError
+	if errors.As(err, &canceledErr) {
 		// handle cancellation
-	case *workflow.TimeoutError:
-		// handle timeout, could check timeout type by err.TimeoutType()
-	case *workflow.PanicError:
-		// handle panic
+	}
+
+	var timeoutErr *TimeoutError
+	if errors.As(err, &timeoutErr) {
+		// handle timeout, could check timeout type by timeoutErr.TimeoutType()
+	}
+
+	var panicErr *PanicError
+	if errors.As(err, &panicErr) {
+		// handle panic, message and stack trace are available by panicErr.Error() and panicErr.StackTrace()
 	}
 }
 
-Errors from child workflow should be handled in a similar way, except that there should be no *PanicError from child workflow.
-When panic happen in workflow implementation code, temporal client library catches that panic and causing the decision timeout.
+Errors from child workflow should be handled in a similar way, except that instance of *ChildWorkflowExecutionError is returned to
+workflow code. It will contains *ActivityTaskError, which in turn will contains on of the errors above.
+When panic happen in workflow implementation code, SDK catches that panic and causing the decision timeout.
 That decision task will be retried at a later time (with exponential backoff retry intervals).
+
+Workflow consumers will get an instance of *WorkflowExecutionError. This error will contains one of errors above.
 */
 
 type (
-	// CustomError returned from workflow and activity implementations with reason and optional details.
-	CustomError struct {
-		reason  string
-		details Values
-	}
-
-	// GenericError returned from workflow/workflow when the implementations return errors other than from NewCustomError() API.
-	GenericError struct {
-		err string
+	// ApplicationError returned from activity implementations with message and optional details.
+	ApplicationError struct {
+		message      string
+		orginalType  string
+		nonRetryable bool
+		details      Values
 	}
 
 	// TimeoutError returned when activity or child workflow timed out.
 	TimeoutError struct {
-		timeoutType commonpb.TimeoutType
-		details     Values
+		timeoutType          commonpb.TimeoutType
+		lastErr              error
```

Comment:
Can we include some documentation here?  
I assume error here could be of following types:
1. ActivityTaskError
2. ChildWorkflowTaskError

Reply:
No, it is actual last error from activity itself (i.e. `ApplicationError` or another `TimeoutError`). I can't add it right now but I will do with in next timeout repated PR.